### DISCUSSION TITLE
Species as action arguments & variable scope in create statements

### DISCUSSION THREAD

**Discussion by DotIN13** [gama-platform/gama#79] [General]
In some certain cases, the variable types can be a little tricky.

For example, when using species declaration names as action arguments to access all agents, it is possible to use either of the following combinations.

```java
species Foo {}

action test_action(species<agent> foo) {}
do test_action(Foo);

action test_action_2(container<agent> foo) {}
do test_action_2(Foo.population);
do test_action_2(Foo);

action test_action_3(list<agent> foo) {}
do test_action_3(list(Foo));
```

And when creating agents, the scope actually shift to the species of craetion as if in an `ask` block, which might cause unexpected shadowing.

```java
species Foo {
    int bar <- 0;
}

species Baz {
    int bar <- 1;
    create Foo number: 1 {
        write bar;
        write myself.bar;
    }
}

// => 0
// => 1
```

This might be already in the documentation, but I just wanted to spin it out here and hope it will help people!

And ps. The algolia of the documentation site is just below the bars of smartness I guess, for example when I search for `create`, I do not get the statement page in the results.

**Comment by benoitgaudou** [comment]
Dear,

Here are some answers/remarks/comments on your 3  points:

*Point 1: ambiguity between species  and population*

*In some certain cases, the variable types can be a little tricky, while
lacking documentation.*

*For example, when using species declaration names as action arguments to
access all agents, it is possible to use either of the following
combinations.*

*species Foo {}

action  test_action(species<agent> foo) {}
do test_action(Foo);

action test_action_2(container<agent> foo) {}
do test_action_2(Foo.population);
do test_action_2(Foo);

action test_action_3(list<agent> foo) {}
do test_action_3(list(Foo));*

Pay attention to the fact that in all these actions, foo is not of the same
type.
My advise:
1) If  you  want to define an action in  which you want to create agents of
a species given  as parameter, choose  the first action  (foo in the
test_action is a  species):

*action* test_action(*species<agent>* foo) {

*create* *foo*;

*write* *foo*;

*write* *list*(*foo*);

}


*reflex* toto {

*do* *test_action*(*Foo*);
                 }
2) If  you want to use the list of all the agents of  a species, use the
last one:

*action* test_action_3(*list<Foo>* foo) {*write* *foo*;}



*reflex* toto {

*do* *test_action_3*(*list*(*Foo*));

*do* *test_action_3*(*Foo* *as* *list*);

}



*Point 2: scope in  the create statement:*

The documentation of the statement indeed makes it explicit that create
behaves as  an ask. I admit the word scope could introduce in the doc in
order to make more explicit.
https://gama-platform.org/wiki/Statements#create

   - If one wants to specify a special initialization sequence for the
   agents created, create provides the same possibilities as ask. This
   extended syntax is:

create a_species number: an_int {
     [statements]
}


   - The same rules as in ask apply. The only difference is that, for the
   agents created, the assignments of variables will bypass the initialization
   defined in species. For instance:

create species(self) number: rnd (4) returns: children {
     set location <- myself.location + {rnd (2), rnd (2)}; // tells
the children to be initially located close to me
     set parent <- myself; // tells the children that their parent is
me (provided the variable parent is declared in this species)
}



*Point 3: documentation of the create  statement:*

*And ps. The algolia of the documentation site is just below the bars of
smartness I guess, for example when I search for create, I do not get the
statement page in the results.*
Even if it is not the first result, the statement page appears in the
results...
And all the other results are perfectly coherent with a search with the
"create" word.

[image: Screenshot 2022-07-15 at 11.41.23.png]

Being more precise with "create statement", the first result is the one
expected.
[image: Screenshot 2022-07-15 at 11.43.38.png]

Do not hesitate to send feedback and proposal on how to improve the
documentation.

Cheers

Benoit

**Reply by DotIN13** [reply]
Thanks for your thoughts!

For the first case, I agree that `species<agent>` and `list<agent>` would be the most coherent representations to use. Regarding this, I had some more interesting ideas to share, maybe tomorrow.

And for the second, I put it here simply because I took the fall when trying to assign values with create. And the fact that I did not read the documentation of create statement put me to debug a *bit* longer than usual <whooosh>. Maybe we could put the head-up in the `Learn GAML step by step - Global Species - Regular species - Instantiate an agent` part,  just like we did in the `The ask statement` part, so it may well become more intuitive.

And it is worth noticing that the init statements of a certain agent with a parent species have a fixed running sequence. First parent `init`, then child `init`, then create statement `init`. And that is probably the reason create statement initialization could bypass species `init`.

As for the documentation and algolia, I think they just kinda needed improvements, however based on my skinny guesses, I wonder if they are generated with java code directly, which I have no knowledge of (XD). Or is there any way I could possibly contribute to the documentation?

**Reply by lesquoyb** [reply]
Hello,

We just changed the priority of pages in the search engine three days ago so that part should be a little bit better now. For the documentation, some is generated from the source code you are right, those pages have the sentence `This file is automatically generated from java files. Do Not Edit It.` written on top of them. 
![image](https://user-images.githubusercontent.com/6374469/179447339-1f1c8dc4-0e33-4808-b467-4dbe548c5b54.png)
Here an example with the Statements page.

But most pages are editable by users, you will find a button `Edit this page` at the bottom of every page.
![image](https://user-images.githubusercontent.com/6374469/179447536-45dbd0e0-6fe6-4da6-89ae-437865fee3b2.png)
It will redirect you to the corresponding `.md` file in the repository. 
Feel free to contribute if you can spot some  lacking parts of the documentation, your changes will be available on the `next` version of the website once it will be rebuilt.

**Reply by lesquoyb** [reply]
Oh, I thought you were part of the gama-platform organization! So you actually don't have the rights necessary to edit the `.md` files, but if you have some modification you want to do or just signal a problem/lack of documentation, you can do it by opening an issue on the [repository dedicated to the website/documentation](https://github.com/gama-platform/gama-platform.github.io/issues)
